ethereum/sidecar/server.go

```diff
@@ -428,6 +540,342 @@ func (s *Server) startGRPCServer(
+	recentEvents, err := s.fetchRecentAssetsUnlockedEvents(ctx)
+	if err != nil {
+		return fmt.Errorf(
+			"failed to fetch recent AssetsUnlocked events: [%w]",
+			err,
+		)
+	}
+
+	unattestedEvents, err := s.findUnattestedAssetsUnlockedEvents(
+		ctx,
+		recentEvents,
+	)
+	if err != nil {
+		return fmt.Errorf(
+			"failed to find unattested AssetsUnlocked events: [%w]",
+			err,
+		)
+	}
+
+	s.attestationMutex.Lock()
+	s.attestationQueue = unattestedEvents
+	s.attestationMutex.Unlock()
+
+	// Save the unlock sequence of the last event as the starting point for
+	// further event fetching.
+	s.lastAssetsUnlockedSequence = sdkmath.ZeroInt()
+	if len(recentEvents) > 0 {
+		s.lastAssetsUnlockedSequence = recentEvents[len(recentEvents)-1].UnlockSequence
+	}
+
+	s.logger.Info(
+		"Initial search for recent unattested AssetsUnlocked events",
+		"recent_events", len(recentEvents),
+		"unattested_events", len(unattestedEvents),
+		"unlock_sequence_tip", s.lastAssetsUnlockedSequence.String(),
+	)
+
+	ticker := time.NewTicker(assetsUnlockedFetchingPeriod)
+	defer ticker.Stop()
+	tickerChan := ticker.C
+
+	for {
+		select {
+		case <-ctx.Done(): // Handle context cancellation
+			s.logger.Info(
+				"stopping assets unlocked event observation due to context cancellation",
+			)
+			return nil
+		case <-tickerChan:
+			err := s.fetchNewAssetsUnlockedEvents(ctx)
+			if err != nil {
+				s.logger.Error(
+					"failed to monitor newly emitted assets unlocked events",
+					"err", err,
+				)
+			}
+		}
+	}
+}
+
+// fetchRecentAssetsUnlockedEvents fetches AssetsUnlocked events that entered
+// the Mezo blockchain within the AssetsUnlocked look-back period. We consider
+// events within that range as possibly unattested. The returned events are
+// sorted in the ascending order of their unlock sequences (from the lowest
+// to the highest).
+func (s *Server) fetchRecentAssetsUnlockedEvents(ctx context.Context) (
+	[]bridgetypes.AssetsUnlockedEvent,
+	error,
+) {
+	// Fetching events starts from the current sequence tip.
+	sequenceTip, err := s.bridgeOutClient.GetAssetsUnlockedSequenceTip(ctx)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"failed to fetch assets unlocked sequence tip: [%w]",
+			err,
+		)
+	}
+
+	// The sequence tip is zero, meaning no events have been made.
+	if sequenceTip.IsZero() {
+		return []bridgetypes.AssetsUnlockedEvent{}, nil
+	}
+
+	// Cut-off block time in UNIX seconds.
+	cutOffBlockTime := uint32(s.timeFunc().Add(-s.bridgeOutLookBackPeriod).Unix()) //nolint:gosec
+	recentEvents := []bridgetypes.AssetsUnlockedEvent{}
+
+	// Walk backwards from the current sequence tip in windows of at most
+	// `assetsUnlockedBatchSize`.
+outer:
+	for sequenceTip.IsPositive() {
+		// Sequence end is exclusive. We must add `1`.
+		seqEnd := sequenceTip.AddRaw(1)
+
+		// Sequence start is inclusive. It cannot be lower than `1`.
+		seqStart := sequenceTip.SubRaw(int64(s.assetsUnlockedBatchSize - 1))
```

Comment:
Would it be simpler to do:
```suggestion
		seqStart := seqEnd.SubRaw(int64(s.assetsUnlockedBatchSize))
```